ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java

```diff
@@ -0,0 +1,344 @@
+      throw new JSApplicationIllegalArgumentException("ANimated node " + animatedNodeTag + " is " +
+        "already attached to a view");
+    }
+    propsAnimatedNode.mConnectedViewTag = viewTag;
+  }
+
+  public void disconnectAnimatedNodeFromView(int animatedNodeTag, int viewTag) {
+    AnimatedNode node = mAnimatedNodes.get(animatedNodeTag);
+    if (node == null) {
+      throw new JSApplicationIllegalArgumentException("Animated node with tag " + animatedNodeTag +
+        " does not exists");
+    }
+    if (!(node instanceof PropsAnimatedNode)) {
+      throw new JSApplicationIllegalArgumentException("Animated node connected to view should be" +
+        "of type " + PropsAnimatedNode.class.getName());
+    }
+    PropsAnimatedNode propsAnimatedNode = (PropsAnimatedNode) node;
+    if (propsAnimatedNode.mConnectedViewTag == viewTag) {
+      propsAnimatedNode.mConnectedViewTag = -1;
+    }
+  }
+
+  /**
+   * Animation loop performs two DFSes over the graph of animated nodes. We use incremented
+   * {@code mAnimatedGraphDFSColor} to mark nodes as visited in each of the DFSes which saves
+   * additional loops for clearing "visited" states.
+   *
+   * First DFS starts with nodes that are in {@code mUpdatedNodes} or directly attached to an active
+   * animation (hence linked to objects from {@code mActiveAnimations}). In that step we calculate
+   * an attribute {@code mActiveIncomingNodes}. The second DFS runs in topological order over the
+   * sub-graph of *active* nodes. This is done by adding node to the DFS queue only if all its
+   * "predecessors" have already been visited.
+   */
+  private void runAnimationStep(long frameTimeNanos) {
+    int activeNodesCount = 0;
+    int updatedNodesCount = 0;
+    boolean hasFinishedAnimations = false;
+
+    // STEP 1.
+    // DFS over graph of nodes starting from ones from `mUpdatedNodes` and ones that are attached to
+    // active animations (from `mActiveANimations`. Update `mIncomingNodes` attribute for each node
+    // during that DFS. Store number of visited nodes in `activeNodesCount`. We "execute" active
+    // animations as a part of this step.
+
+    mAnimatedGraphDFSColor++; /* use new color */
+    if (mAnimatedGraphDFSColor == AnimatedNode.INITIAL_DFS_COLOR) {
+      // value "0" is used as an initial color for a new node, using it in DFS may cause some nodes
+      // to be skipped.
+      mAnimatedGraphDFSColor++;
+    }
+
+    Queue<AnimatedNode> nodesQueue = new ArrayDeque<>();
+    for (int i = 0; i < mUpdatedNodes.size(); i++) {
+      AnimatedNode node = mUpdatedNodes.get(i);
+      if (node.mDFSColor != mAnimatedGraphDFSColor) {
+        node.mDFSColor = mAnimatedGraphDFSColor;
+        activeNodesCount++;
+        nodesQueue.add(node);
+      }
+    }
+
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      animation.runAnimationStep(frameTimeNanos);
```

Comment:
goo catch. The return value was used to indicate if the value has changed. After refactoring I'm no longer using that information (I though it could've been used for optimising traversal but I believe it's not really worth doing as most of the animated graphs aren't very huge)